# Importing Data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df=pd.read_csv('Iowa_Liquor_Sales.csv',delimiter=',', low_memory=False)

In [3]:
df.head()

Invoice/Item Number        Date  Store Number  \
0        S29198800001  11/20/2015          2191   
1        S29195400002  11/21/2015          2205   
2        S29050300001  11/16/2015          3549   
3        S28867700001  11/04/2015          2513   
4        S29050800001  11/17/2015          3942   

                         Store Name              Address          City  \
0                    Keokuk Spirits            1013 MAIN        KEOKUK   
1            Ding's Honk And Holler     900 E WASHINGTON      CLARINDA   
2              Quicker Liquor Store         1414 48TH ST  FORT MADISON   
3  Hy-Vee Food Store #2 / Iowa City       812  S 1ST AVE     IOWA CITY   
4                  Twin Town Liquor  104 HIGHWAY 30 WEST        TOLEDO   

  Zip Code                                     Store Location  County Number  \
0    52632    1013 MAIN\nKEOKUK 52632\n(40.39978, -91.387531)           56.0   
1    51632  900 E WASHINGTON\nCLARINDA 51632\n(40.739238, ...           73.0   
2    52627  1414 48TH ST\nFORT MADISON 52627\n(40.624226, ...           56.0   
3    52240                   812 S 1ST AVE\nIOWA CITY 52240\n           52.0   
4    52342  104 HIGHWAY 30 WEST\nTOLEDO 52342\n(41.985887,...           86.0   

    County  ...  Item Number                               Item Description  \
0      Lee  ...          297                          Templeton Rye w/Flask   
1     Page  ...          297                          Templeton Rye w/Flask   
2      Lee  ...          249  Disaronno Amaretto Cavalli Mignon 3-50ml Pack   
3  Johnson  ...          237                 Knob Creek w/ Crystal Decanter   
4     Tama  ...          249  Disaronno Amaretto Cavalli Mignon 3-50ml Pack   

   Pack Bottle Volume (ml)  State Bottle Cost State Bottle Retail  \
0     6                750             $18.09              $27.14   
1     6                750             $18.09              $27.14   
2    20                150              $6.40               $9.60   
3     3               1750             $35.55              $53.34   
4    20                150              $6.40               $9.60   

   Bottles Sold  Sale (Dollars) Volume Sold (Liters) Volume Sold (Gallons)  
0             6         $162.84                 4.50                  1.19  
1            12         $325.68                 9.00                  2.38  
2             2          $19.20                 0.30                  0.08  
3             3         $160.02                 5.25                  1.39  
4             2          $19.20                 0.30                  0.08  

[5 rows x 24 columns]

In [5]:
df.shape

(12591077, 24)

# Exploring the Data

In [6]:
df.dropna(inplace=True)
df.shape

(12495974, 24)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12495974 entries, 6 to 12591076
Data columns (total 24 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Invoice/Item Number    object 
 1   Date                   object 
 2   Store Number           int64  
 3   Store Name             object 
 4   Address                object 
 5   City                   object 
 6   Zip Code               object 
 7   Store Location         object 
 8   County Number          float64
 9   County                 object 
 10  Category               float64
 11  Category Name          object 
 12  Vendor Number          float64
 13  Vendor Name            object 
 14  Item Number            int64  
 15  Item Description       object 
 16  Pack                   int64  
 17  Bottle Volume (ml)     int64  
 18  State Bottle Cost      object 
 19  State Bottle Retail    object 
 20  Bottles Sold           int64  
 21  Sale (Dollars)         object 
 22  Volume Sold (Lit

In [9]:
df['Date'] = pd.to_datetime(df['Date'])

In [13]:
df['Sale (Dollars)'] = df['Sale (Dollars)'].str.replace('$', '')
df['Sale (Dollars)'] = df['Sale (Dollars)'].astype('float')

df['State Bottle Cost'] = df['State Bottle Cost'].str.replace('$', '')
df['State Bottle Cost'] = df['State Bottle Cost'].astype('float')

df['State Bottle Retail'] = df['State Bottle Retail'].str.replace('$', '')
df['State Bottle Retail'] = df['State Bottle Retail'].astype('float')

<ipython-input-13-fd9075865ee9>:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['State Bottle Cost'] = df['State Bottle Cost'].str.replace('$', '')
<ipython-input-13-fd9075865ee9>:7: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['State Bottle Retail'] = df['State Bottle Retail'].str.replace('$', '')


In [14]:
df.dtypes

Invoice/Item Number              object
Date                     datetime64[ns]
Store Number                      int64
Store Name                       object
Address                          object
City                             object
Zip Code                         object
Store Location                   object
County Number                   float64
County                           object
Category                        float64
Category Name                    object
Vendor Number                   float64
Vendor Name                      object
Item Number                       int64
Item Description                 object
Pack                              int64
Bottle Volume (ml)                int64
State Bottle Cost               float64
State Bottle Retail             float64
Bottles Sold                      int64
Sale (Dollars)                  float64
Volume Sold (Liters)            float64
Volume Sold (Gallons)           float64
dtype: object

In [15]:
df.head()

Invoice/Item Number       Date  Store Number  \
6         S28865700001 2015-11-09          2538   
8         S29339300091 2015-11-30          2662   
13        S28866900001 2015-11-11          3650   
18        S29134300126 2015-11-18          3723   
21        S29282800048 2015-11-23          2642   

                           Store Name                Address       City  \
6     Hy-Vee Food Store #3 / Waterloo       1422 FLAMMANG DR   WATERLOO   
8   Hy-Vee Wine & Spirits / Muscatine  522 MULBERRY, SUITE A  MUSCATINE   
13         Spirits, Stogies and Stuff     118 South Main St.   HOLSTEIN   
18                 J D Spirits Liquor           1023  9TH ST      ONAWA   
21    Hy-Vee Wine and Spirits / Pella        512 E OSKALOOSA      PELLA   

   Zip Code                                     Store Location  County Number  \
6     50702  1422 FLAMMANG DR\nWATERLOO 50702\n(42.459938, ...            7.0   
8     52761           522 MULBERRY, SUITE A\nMUSCATINE 52761\n           70.0   
13    51025  118 South Main St.\nHOLSTEIN 51025\n(42.490073...           47.0   
18    51040  1023 9TH ST\nONAWA 51040\n(42.025841, -96.095845)           67.0   
21    50219  512 E OSKALOOSA\nPELLA 50219\n(41.397023, -92....           63.0   

        County  ...  Item Number              Item Description  Pack  \
6   Black Hawk  ...          238  Forbidden Secret Coffee Pack     6   
8    Muscatine  ...          173   Laphroaig w/ Whiskey Stones    12   
13         Ida  ...          238  Forbidden Secret Coffee Pack     6   
18      Monona  ...          258           Rumchata "GoChatas"     1   
21      Marion  ...          238  Forbidden Secret Coffee Pack     6   

   Bottle Volume (ml)  State Bottle Cost State Bottle Retail  Bottles Sold  \
6                1500              11.62               17.43             6   
8                 750              19.58               29.37             4   
13               1500              11.62               17.43             1   
18               6000              99.00              148.50             1   
21               1500              11.62               17.43             6   

    Sale (Dollars)  Volume Sold (Liters)  Volume Sold (Gallons)  
6           104.58                   9.0                   2.38  
8           117.48                   3.0                   0.79  
13           17.43                   1.5                   0.40  
18          148.50                   6.0                   1.59  
21          104.58                   9.0                   2.38  

[5 rows x 24 columns]

In [16]:
df['Store Number'] = df['Store Number'].astype('str')
df['Vendor Number'] = df['Vendor Number'].astype('str')

In [17]:
df['State Bottle Profit'] = df['State Bottle Retail'] - df['State Bottle Cost']

# Data Analysis and Data Visualization

In [18]:
df.nunique()      

Invoice/Item Number      12495974
Date                         1379
Store Number                 1882
Store Name                   1948
Address                      3111
City                          790
Zip Code                      450
Store Location               4405
County Number                  99
County                        200
Category                      104
Category Name                 130
Vendor Number                 265
Vendor Name                   384
Item Number                  7263
Item Description             5744
Pack                           28
Bottle Volume (ml)             49
State Bottle Cost            2727
State Bottle Retail          3183
Bottles Sold                  448
Sale (Dollars)              23056
Volume Sold (Liters)         1174
Volume Sold (Gallons)        1180
State Bottle Profit          2565
dtype: int64

In [19]:
df['City'].value_counts().head(5)

DES MOINES      809000
CEDAR RAPIDS    595436
DAVENPORT       409334
Des Moines      300183
WATERLOO        288771
Name: City, dtype: int64

That means that there are the highest number of shops in DES MOINES from 730 unique cities in our data.

In [20]:
df['Category Name'].value_counts().head(5)

VODKA 80 PROOF               1255745
CANADIAN WHISKIES             928346
STRAIGHT BOURBON WHISKIES     539875
SPICED RUM                    526385
VODKA FLAVORED                498464
Name: Category Name, dtype: int64

The most popular category name is VODKA 80 PROOF from 130 unique categories in our data.

###  Sales of liqour per city in dollars 

In [35]:
import plotly.express as px
import plotly as py
import plotly.graph_objects as go

df_city = df.groupby(['City','Pack','Date'])['Sale (Dollars)'].sum().groupby(['City','Pack']).max().sort_values().groupby(['City']).sum().sort_values(ascending=False)
top_city = pd.DataFrame(df_city)
top_city = pd.DataFrame(df_city.head(20))

fig_reg = px.bar(top_city,x=top_city.index, y='Sale (Dollars)',color='Sale (Dollars)')
fig_reg.update_layout(
    title="Sales of liquor per city",
    xaxis_title=" City ",
    yaxis_title="Sales in dollars",
    )
fig_reg.show()

### The least sales of liqour per city in dollars 

In [42]:
df_city1 = df.groupby(['City','Pack','Date'])['Sale (Dollars)'].sum().groupby(['City','Pack']).max().sort_values().groupby(['City']).sum().sort_values(ascending=True)
top_city1 = pd.DataFrame(df_city1)
top_city1 = pd.DataFrame(df_city1.head(20))

fig_reg = px.bar(top_city1,x=top_city1.index, y='Sale (Dollars)',color='Sale (Dollars)')
fig_reg.update_layout(
    title="The least sales of liquor per city",
    xaxis_title=" City ",
    yaxis_title="Sales in dollars",
    )
fig_reg.show()

### Sales of liqour per store in dollars 

In [36]:
df_store = df.groupby(['Store Name','Pack','Date'])['Sale (Dollars)'].sum().groupby(['Store Name','Pack']).max().sort_values().groupby(['Store Name']).sum().sort_values(ascending=False)
top_store = pd.DataFrame(df_store)
top_store = pd.DataFrame(df_store.head(20))

fig_reg = px.bar(top_store,x=top_store.index, y='Sale (Dollars)',color='Sale (Dollars)')
fig_reg.update_layout(
    title="Sales of liquor per store",
    xaxis_title=" Store Name ",
    yaxis_title="Sales in dollars",
    )
fig_reg.show()

### The least sales of liqour per store in dollars 

In [43]:
df_store1 = df.groupby(['Store Name','Pack','Date'])['Sale (Dollars)'].sum().groupby(['Store Name','Pack']).max().sort_values().groupby(['Store Name']).sum().sort_values(ascending=True)
top_store1 = pd.DataFrame(df_store1)
top_store1 = pd.DataFrame(df_store1.head(20))

fig_reg = px.bar(top_store1,x=top_store1.index, y='Sale (Dollars)',color='Sale (Dollars)')
fig_reg.update_layout(
    title="Sales of liquor per store",
    xaxis_title=" Store Name ",
    yaxis_title="Sales in dollars",
    )
fig_reg.show()

### Sales of liqour per category

In [38]:
df_name = df.groupby(['Category Name','Pack','Date'])['Sale (Dollars)'].sum().groupby(['Category Name','Pack']).max().sort_values().groupby(['Category Name']).sum().sort_values(ascending=False)

top_name = pd.DataFrame(df_name)
top_name = pd.DataFrame(df_name.head(20))

fig_reg = px.bar(top_name,x=top_name.index, y='Sale (Dollars)',color='Sale (Dollars)')
fig_reg.update_layout(
    title="Sales of liquor per category",
    xaxis_title=" Category Name",
    yaxis_title="Sales in dollars",
    )
fig_reg.show()

### The least sales of liqour per category

In [44]:
df_name1 = df.groupby(['Category Name','Pack','Date'])['Sale (Dollars)'].sum().groupby(['Category Name','Pack']).max().sort_values().groupby(['Category Name']).sum().sort_values(ascending=True)

top_name1 = pd.DataFrame(df_name1)
top_name1 = pd.DataFrame(df_name1.head(20))

fig_reg = px.bar(top_name1,x=top_name1.index, y='Sale (Dollars)',color='Sale (Dollars)')
fig_reg.update_layout(
    title="Sales of liquor per category",
    xaxis_title=" Category Name",
    yaxis_title="Sales in dollars",
    )
fig_reg.show()

### State Bottle Profit in dollars 

In [40]:
df_bottle = df.groupby(['Category Name','Pack','Date'])['State Bottle Profit'].sum().groupby(['Category Name','Pack']).max().sort_values().groupby(['Category Name']).sum().sort_values(ascending=False)

top_bottle = pd.DataFrame(df_bottle)
top_bottle = pd.DataFrame(df_bottle.head(20))

fig_reg = px.bar(top_bottle,x=top_bottle.index, y='State Bottle Profit',color='State Bottle Profit')
fig_reg.update_layout(
    title="State Bottle Profit in dollars",
    xaxis_title=" Category Name",
    yaxis_title="State Bottle Profit in dollars",
    )
fig_reg.show()

### The least State Bottle Profit in dollars 

In [45]:
df_bottle1 = df.groupby(['Category Name','Pack','Date'])['State Bottle Profit'].sum().groupby(['Category Name','Pack']).max().sort_values().groupby(['Category Name']).sum().sort_values(ascending=True)

top_bottle1 = pd.DataFrame(df_bottle1)
top_bottle1 = pd.DataFrame(df_bottle1.head(20))

fig_reg = px.bar(top_bottle1,x=top_bottle1.index, y='State Bottle Profit',color='State Bottle Profit')
fig_reg.update_layout(
    title="State Bottle Profit in dollars",
    xaxis_title=" Category Name",
    yaxis_title="State Bottle Profit in dollars",
    )
fig_reg.show()

### State Bottle Profit per city in dollars 

In [46]:
df_bottle_city = df.groupby(['City','Pack','Date'])['State Bottle Profit'].sum().groupby(['City','Pack']).max().sort_values().groupby(['City']).sum().sort_values(ascending=False)

top_bottle_city = pd.DataFrame(df_bottle_city)
top_bottle_city = pd.DataFrame(df_bottle_city.head(20))

fig_reg = px.bar(top_bottle_city,x=top_bottle_city.index, y='State Bottle Profit',color='State Bottle Profit')
fig_reg.update_layout(
    title="State Bottle Profit per city in dollars",
    xaxis_title=" City",
    yaxis_title="State Bottle Profit in dollars",
    )
fig_reg.show()

### The least State Bottle Profit per city in dollars 

In [48]:
df_bottle_city1 = df.groupby(['City','Pack','Date'])['State Bottle Profit'].sum().groupby(['City','Pack']).max().sort_values().groupby(['City']).sum().sort_values(ascending=True)

top_bottle_city1 = pd.DataFrame(df_bottle_city1)
top_bottle_city1 = pd.DataFrame(df_bottle_city1.head(20))

fig_reg = px.bar(top_bottle_city1,x=top_bottle_city1.index, y='State Bottle Profit',color='State Bottle Profit')
fig_reg.update_layout(
    title="State Bottle Profit per city in dollars",
    xaxis_title=" City",
    yaxis_title="State Bottle Profit in dollars",
    )
fig_reg.show()

### Sales per day in dollars 

In [51]:
sales_per_day = df.groupby('Date', as_index=False)['Sale (Dollars)'].sum()
sales_per_day_sc = go.Scatter(x=sales_per_day['Date'], y=sales_per_day['Sale (Dollars)'])
layout = go.Layout(title='Daily sales', xaxis=dict(title='Date'), yaxis=dict(title='Sales'))
fig = go.Figure(data=[sales_per_day_sc], layout=layout)
fig.show()